### PR TITLE
Updated documentation.

### PR DESCRIPTION
**Pull request by mfateev**

### CODE REVIEW

File: src/main/java/com/uber/cadence/client/WorkflowClient.java [hunk 1]
```diff
@@ -31,43 +31,74 @@
...
+ * WorkflowClient workflowClient = WorkflowClient.newClient(cadenceServiceHost, cadenceServicePort, domain);
+ * // Create a workflow stub.
+ * FileProcessingWorkflow workflow = workflowClient.newWorkflowStub(FileProcessingWorkflow.class);
+ * </code></pre>
  *
- * // Create client side stub to the workflow execution.
- * HelloWorldWorkflow workflow = client.newWorkflowStub(HelloWorldWorkflow.class);
+ * There are two ways to start workflow execution: synchronously and asynchronously. Synchronous
+ * invocation starts a workflow and then waits for its completion. If the process that started the
+ * workflow crashes or stops the waiting, the workflow continues executing. Because workflows are
```
Comment (purplesusan) [inline src/main/java/com/uber/cadence/client/WorkflowClient.java]: stops the waiting -> stops waiting

File: src/main/java/com/uber/cadence/client/WorkflowClient.java [hunk 2]
```diff
@@ -31,43 +31,74 @@
...
- * // Start Workflow Execution
- * WorkflowExecution started = WorkflowClient.start(workflow::helloWorld, "User");
+ * <p>Synchronous start:
  *
- * // started.getWorkflowId() should match the one in the options: "MyHelloWorld1"
- * System.out.println("Started helloWorld workflow with workflowId=\"" + started.getWorkflowId()
- *                     + "\" and runId=\"" + started.getRunId() + "\"");
- * </pre>
+ * <pre><code>
+ * // Start a workflow and the wait for a result.
```
Comment (purplesusan) [inline src/main/java/com/uber/cadence/client/WorkflowClient.java]: and the wait -> and wait

File: src/main/java/com/uber/cadence/client/WorkflowClient.java [hunk 3]
```diff
@@ -31,43 +31,74 @@
...
+ * // Note that if the waiting process is killed, the workflow will continue execution.
```
Comment (purplesusan) [inline src/main/java/com/uber/cadence/client/WorkflowClient.java]: execution -> executing.

File: src/main/java/com/uber/cadence/client/WorkflowClient.java [hunk 4]
```diff
@@ -31,43 +31,74 @@
...
- * <pre>
+ * Asynchronous when the result is needed:
+ *
+ * <pre><code>
  * CompletableFuture&lt;String&gt; result = WorkflowClient.execute(workflow::helloWorld, "User");
- * </pre>
+ * </code></pre>
+ *
+ * If you need to wait for a workflow completion after an asynchronous start, may be even from a
+ * different process the simplest way is to call the blocking version again. If {@link
```
Comment (purplesusan) [inline src/main/java/com/uber/cadence/client/WorkflowClient.java]: add a comma after process

File: src/main/java/com/uber/cadence/client/WorkflowClient.java [hunk 5]
```diff
@@ -31,43 +31,74 @@
...
  *
- * <pre>
+ * Asynchronous when the result is needed:
+ *
+ * <pre><code>
  * CompletableFuture&lt;String&gt; result = WorkflowClient.execute(workflow::helloWorld, "User");
- * </pre>
+ * </code></pre>
+ *
+ * If you need to wait for a workflow completion after an asynchronous start, may be even from a
```
Comment (purplesusan) [inline src/main/java/com/uber/cadence/client/WorkflowClient.java]: may be -> maybe

File: src/main/java/com/uber/cadence/common/MethodRetry.java [hunk 6]
```diff
@@ -28,17 +28,18 @@
  * Indicates retry policy for a workflow or activity method. This annotation applies only to
```
Comment (purplesusan) [inline src/main/java/com/uber/cadence/common/MethodRetry.java]: Indicates retry policy -> Indicates a retry policy

File: src/main/java/com/uber/cadence/internal/sync/ActivityTaskImpl.java [hunk 7]
```diff
@@ -45,36 +46,31 @@ public String getActivityId() {
...
-  public ActivityType getActivityType() {
-    return response.getActivityType();
+  public String getActivityType() {
+    return response.getActivityType().getName();
   }
 
   @Override
   public long getScheduledTimestamp() {
-    return response.getScheduledTimestamp();
+    // Cadence timestamp is in microseconds
```
Comment (purplesusan) [inline src/main/java/com/uber/cadence/internal/sync/ActivityTaskImpl.java]: pls add a period at the end.

File: src/main/java/com/uber/cadence/testing/TestActivityEnvironment.java [hunk 8]
```diff
@@ -30,10 +59,30 @@ static TestActivityEnvironment newInstance(TestEnvironmentOptions options) {
     return new TestActivityEnvironmentInternal(options);
   }
 
+  /**
+   * Registers activities implementations to test. Use {@link #newActivityStub(Class)} to create
```
Comment (purplesusan) [inline src/main/java/com/uber/cadence/testing/TestActivityEnvironment.java]: activities -> activity

File: src/main/java/com/uber/cadence/testing/TestWorkflowEnvironment.java [hunk 9]
```diff
@@ -27,13 +28,13 @@
...
- * <p>Testing workflow code is hard as it might be potentially very long running. The included
- * in-memory implementation of the Cadence service supports <b>automatic time skipping</b>. Anytime
- * a workflow under the test as well as a unit test code are waiting on a timer (or sleep) the
- * internal service time is automatically advanced to the nearest time that unblocks one of the
- * waiting threads. This way a workflow that runs in production for months is unit tested in
+ * <p>Testing the workflow code is hard as it might be potentially very long running. The included
+ * in-memory implementation of the Cadence service supports <b>an automatic time skipping</b>.
+ * Anytime a workflow under the test as well as the unit test code are waiting on a timer (or sleep)
+ * the internal service time is automatically advanced to the nearest time that unblocks one of the
+ * waiting threads. This way a workflow that runs in the production for months is unit tested in
```
Comment (purplesusan) [inline src/main/java/com/uber/cadence/testing/TestWorkflowEnvironment.java]: runs in the production -> runs in production OR runs in the production environment

File: src/main/java/com/uber/cadence/workflow/Workflow.java [hunk 10]
```diff
@@ -21,11 +21,342 @@
...
+import com.uber.cadence.workflow.Functions.Func2;
+import com.uber.cadence.workflow.Functions.Proc;
 import java.time.Duration;
 import java.util.Optional;
+import java.util.UUID;
 import java.util.concurrent.CancellationException;
 import java.util.function.Supplier;
 
+/**
+ * Workflow encapsulates the orchestration of activities and child workflows. It can also answer to
```
Comment (purplesusan) [inline src/main/java/com/uber/cadence/workflow/Workflow.java]: These comments are all in the readme, which I'm breaking down into docs. Is there a reason the information is repeated here? Let me know if I need to review this file also.

Comment (mfateev) [reply src/main/java/com/uber/cadence/workflow/Workflow.java]: These will end up in the generated JavaDoc. Java developers frequently avoid documentation and go directly to JavaDoc. I thought about removing them from the readme, but as JavaDoc is not very well organized having a separate well organized documentation makes sense. So I think we end up duplicating some of it.

**Pull request merged**